### ISSUE TITLE
ETL cache of hg.mozilla.org is broken

### ISSUE THREAD

**Issue by klahnakoski** [klahnakoski/TestLog-ETL#44]
The ES query delivers too much data, Python runs out of memory, but recovers, and then proceeds to ask hg for the same.  This is repeated by all ETL machines, and too many hg requests.

    ERROR: Problem with search (path=/repo20180418_092942/revision/_search):
	{
	    "query": {"bool": {"must": [
	                {"term": {"changeset.id12": "da28b92efe6f"}},
	                {"term": {"branch.name": "maple"}},
	                {"term": {"branch.locale": "en-US"}},
	                {"range": {"etl.timestamp": {"gt": 1525305600}}}
	            ]}},
	    "size": 2000
	}

**Comment by klahnakoski** [comment]
No, it is not a too-much-data problem; I got that mixed up with some other problem.

The first occurrence is here:

     ip-172-31-2-165 (pid 10486) - 2018-06-28 22:25:28 - hg daemon - "hg_mozilla_org.py:280" (_get_from_elasticsearch) - WARNING: Bad ES call, fall back to HG

**Comment by klahnakoski** [comment]
the node used for HG cache started acting different at  June 29 19:42 UTC and June 29 15:51 UTC

**Comment by klahnakoski** [comment]
2018-06-29 15:32:26 failures became more intense (as more ETL machines came online?)

**Comment by klahnakoski** [comment]
There is more than one hg problem. The one above is a utf8 conversion problem.  There is a "Connection refused'" type starting at 2018-06-29 15:32:17, which better aligns with the server acting different, and explains the volume of requests redirected to HG.

 
```
ERROR: HTTPConnectionPool(host='172.31.1.12', port=9200): Max retries exceeded with url: /repo20180418_092942/revision/_search (Caused by NewConnectionError('<urllib3.connection.HTTPConnection object at 0x7f2e27ca3510>: Failed to establish a new connection: [Errno 111] Connection refused',))ip-172-31-2-78 (pid 19887) - 2018-06-29 15:32:17 - ETL Loop 0 - "hg_mozilla_org.py:280" (_get_from_elasticsearch) - WARNING: Bad ES call, fall back to HG
	File "/home/ubuntu/ActiveData-ETL/vendor/mo_hg/hg_mozilla_org.py", line 280, in _get_from_elasticsearch
	File "/home/ubuntu/ActiveData-ETL/vendor/mo_hg/hg_mozilla_org.py", line 182, in get_revision
	File "/home/ubuntu/ActiveData-ETL/vendor/pyLibrary/meta.py", line 133, in <lambda>
	File "/home/ubuntu/ActiveData-ETL/vendor/pyLibrary/meta.py", line 169, in output
	File "/home/ubuntu/ActiveData-ETL/activedata_etl/transforms/pulse_block_to_task_cluster.py", line 512, in set_build_info
	File "/home/ubuntu/ActiveData-ETL/activedata_etl/transforms/pulse_block_to_task_cluster.py", line 287, in _normalize
	File "/home/ubuntu/ActiveData-ETL/activedata_etl/transforms/pulse_block_to_task_cluster.py", line 96, in process
	File "activedata_etl/etl.py", line 177, in _dispatch_work
	File "activedata_etl/etl.py", line 304, in loop
	File "/home/ubuntu/ActiveData-ETL/vendor/mo_threads/threads.py", line 260, in _run
caused by
	ERROR: Problem with search (path=/repo20180418_092942/revision/_search):
	{
	    "query": {"bool": {"must": [
	                {"term": {"changeset.id12": "99371dbf5392"}},
	                {"term": {"branch.name": "mozilla-inbound"}},
	                {"term": {"branch.locale": "en-US"}},
	                {"range": {"etl.timestamp": {"gt": 1525305600}}}
	            ]}},
	    "size": 2000
	}
	File "/home/ubuntu/ActiveData-ETL/vendor/pyLibrary/env/elasticsearch.py", line 465, in search
	File "/home/ubuntu/ActiveData-ETL/vendor/mo_hg/hg_mozilla_org.py", line 268, in _get_from_elasticsearch
	File "/home/ubuntu/ActiveData-ETL/vendor/mo_hg/hg_mozilla_org.py", line 182, in get_revision
	File "/home/ubuntu/ActiveData-ETL/vendor/pyLibrary/meta.py", line 133, in <lambda>
	File "/home/ubuntu/ActiveData-ETL/vendor/pyLibrary/meta.py", line 169, in output
	File "/home/ubuntu/ActiveData-ETL/activedata_etl/transforms/pulse_block_to_task_cluster.py", line 512, in set_build_info
	File "/home/ubuntu/ActiveData-ETL/activedata_etl/transforms/pulse_block_to_task_cluster.py", line 287, in _normalize
	File "/home/ubuntu/ActiveData-ETL/activedata_etl/transforms/pulse_block_to_task_cluster.py", line 96, in process
	File "activedata_etl/etl.py", line 177, in _dispatch_work
	File "activedata_etl/etl.py", line 304, in loop
	File "/home/ubuntu/ActiveData-ETL/vendor/mo_threads/threads.py", line 260, in _run
caused by
	ERROR: Problem with call to http://172.31.1.12:9200/repo20180418_092942/revision/_search
{"query":{"bool":{"must":[{"term":{"changeset.id12":"99371dbf5392"}},{"term":{"branch.name":"mozilla-inbound"}},{"term":{"branch.locale":"en-US"}},{"range":{"etl.timestamp":{"gt":1525305600}}}]}},"size":2000}
	File "/home/ubuntu/ActiveData-ETL/vendor/pyLibrary/env/elasticsearch.py", line 908, in post
	File "/home/ubuntu/ActiveData-ETL/vendor/pyLibrary/env/elasticsearch.py", line 458, in search
	File "/home/ubuntu/ActiveData-ETL/vendor/mo_hg/hg_mozilla_org.py", line 268, in _get_from_elasticsearch
	File "/home/ubuntu/ActiveData-ETL/vendor/mo_hg/hg_mozilla_org.py", line 182, in get_revision
	File "/home/ubuntu/ActiveData-ETL/vendor/pyLibrary/meta.py", line 133, in <lambda>
	File "/home/ubuntu/ActiveData-ETL/vendor/pyLibrary/meta.py", line 169, in output
	File "/home/ubuntu/ActiveData-ETL/activedata_etl/transforms/pulse_block_to_task_cluster.py", line 512, in set_build_info
	File "/home/ubuntu/ActiveData-ETL/activedata_etl/transforms/pulse_block_to_task_cluster.py", line 287, in _normalize
	File "/home/ubuntu/ActiveData-ETL/activedata_etl/transforms/pulse_block_to_task_cluster.py", line 96, in process
	File "activedata_etl/etl.py", line 177, in _dispatch_work
	File "activedata_etl/etl.py", line 304, in loop
	File "/home/ubuntu/ActiveData-ETL/vendor/mo_threads/threads.py", line 260, in _run
caused by
	ERROR: Tried 1 times: Request failure of http://172.31.1.12:9200/repo20180418_092942/revision/_search
	File "/home/ubuntu/ActiveData-ETL/vendor/pyLibrary/env/http.py", line 171, in request
	File "/home/ubuntu/ActiveData-ETL/vendor/pyLibrary/env/http.py", line 220, in post
	File "/home/ubuntu/ActiveData-ETL/vendor/pyLibrary/env/elasticsearch.py", line 882, in post
	File "/home/ubuntu/ActiveData-ETL/vendor/pyLibrary/env/elasticsearch.py", line 458, in search
	File "/home/ubuntu/ActiveData-ETL/vendor/mo_hg/hg_mozilla_org.py", line 268, in _get_from_elasticsearch
	File "/home/ubuntu/ActiveData-ETL/vendor/mo_hg/hg_mozilla_org.py", line 182, in get_revision
	File "/home/ubuntu/ActiveData-ETL/vendor/pyLibrary/meta.py", line 133, in <lambda>
	File "/home/ubuntu/ActiveData-ETL/vendor/pyLibrary/meta.py", line 169, in output
	File "/home/ubuntu/ActiveData-ETL/activedata_etl/transforms/pulse_block_to_task_cluster.py", line 512, in set_build_info
	File "/home/ubuntu/ActiveData-ETL/activedata_etl/transforms/pulse_block_to_task_cluster.py", line 287, in _normalize
	File "/home/ubuntu/ActiveData-ETL/activedata_etl/transforms/pulse_block_to_task_cluster.py", line 96, in process
	File "activedata_etl/etl.py", line 177, in _dispatch_work
	File "activedata_etl/etl.py", line 304, in loop
	File "/home/ubuntu/ActiveData-ETL/vendor/mo_threads/threads.py", line 260, in _run
caused by
	ERROR: HTTPConnectionPool(host='172.31.1.12', port=9200): Max retries exceeded with url: /repo20180418_092942/revision/_search (Caused by NewConnectionError('<urllib3.connection.HTTPConnection object at 0x7f2e27ca3510>: Failed to establish a new connection: [Errno 111] Connection refused',))
	File "/usr/local/lib/python2.7/dist-packages/requests/adapters.py", line 513, in send
	File "/usr/local/lib/python2.7/dist-packages/requests/sessions.py", line 622, in send
	File "/usr/local/lib/python2.7/dist-packages/requests/sessions.py", line 512, in request
	File "/home/ubuntu/ActiveData-ETL/vendor/pyLibrary/env/http.py", line 164, in request
	File "/home/ubuntu/ActiveData-ETL/vendor/pyLibrary/env/http.py", line 220, in post
	File "/home/ubuntu/ActiveData-ETL/vendor/pyLibrary/env/elasticsearch.py", line 882, in post
	File "/home/ubuntu/ActiveData-ETL/vendor/pyLibrary/env/elasticsearch.py", line 458, in search
	File "/home/ubuntu/ActiveData-ETL/vendor/mo_hg/hg_mozilla_org.py", line 268, in _get_from_elasticsearch
	File "/home/ubuntu/ActiveData-ETL/vendor/mo_hg/hg_mozilla_org.py", line 182, in get_revision
	File "/home/ubuntu/ActiveData-ETL/vendor/pyLibrary/meta.py", line 133, in <lambda>
	File "/home/ubuntu/ActiveData-ETL/vendor/pyLibrary/meta.py", line 169, in output
	File "/home/ubuntu/ActiveData-ETL/activedata_etl/transforms/pulse_block_to_task_cluster.py", line 512, in set_build_info
	File "/home/ubuntu/ActiveData-ETL/activedata_etl/transforms/pulse_block_to_task_cluster.py", line 287, in _normalize
	File "/home/ubuntu/ActiveData-ETL/activedata_etl/transforms/pulse_block_to_task_cluster.py", line 96, in process
	File "activedata_etl/etl.py", line 177, in _dispatch_work
	File "activedata_etl/etl.py", line 304, in loop
	File "/home/ubuntu/ActiveData-ETL/vendor/mo_threads/threads.py", line 260, in _run
```

**Comment by klahnakoski** [comment]
Summary of problem, and solutions

The ETL pipeline consists of many machines that use hg.mozilla.org to lookup repository changesets. Each machine checks a local Elasticsearch (ES) cache for repository entries before going to HG. When an HG call is made, the result is put into the cache for another machine to find later.

In this particular situation, the ES cluster was down; each ETL machine attempted to check the cache; nothing was found; and a request to HG was made. This exposed two problems: 

1. The ETL machines were not checking **why** nothing could be found. This has been fixed; if an ETL machine can not make a connection to ES, then wait 10minutes before continuing.  This serves to slow down the request rate until I have time to fix whatever is causing the problem.
https://github.com/klahnakoski/ActiveData-ETL/commit/831f68d1acb667f7e55c822ff45612cc96db5f8f

2. I did not rate limit all the code paths last time (bug 1417720). The last problem had a similar cause (not identifying a failure mode) on a different code path. This fix apply a rough rate limit on all HG requests (one every 30seconds).
https://github.com/klahnakoski/ActiveData-ETL/commit/115f94ef8642b1a5b7eb4029a6497ccfa2f68cf8
